src/cpu/o3/bac.cc

```diff
@@ -0,0 +1,1069 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "cpu/o3/bac.hh"
+
+#include <algorithm>
+
+#include "arch/generic/pcstate.hh"
+#include "base/trace.hh"
+#include "cpu/inst_seq.hh"
+#include "cpu/o3/dyn_inst.hh"
+#include "cpu/o3/ftq.hh"
+#include "cpu/o3/limits.hh"
+#include "debug/Activity.hh"
+#include "debug/BAC.hh"
+#include "debug/Branch.hh"
+#include "debug/Drain.hh"
+#include "debug/Fetch.hh"
+#include "debug/O3PipeView.hh"
+#include "params/BaseO3CPU.hh"
+#include "sim/full_system.hh"
+
+using namespace gem5::branch_prediction;
+
+namespace gem5
+{
+
+namespace o3
+{
+
+BAC::BAC(CPU *_cpu, const BaseO3CPUParams &params)
+    : cpu(_cpu),
+      bpu(params.branchPred),
+      ftq(nullptr),
+      wroteToTimeBuffer(false),
+      decoupledFrontEnd(params.decoupledFrontEnd),
+      fetchToBacDelay(params.fetchToBacDelay),
+      decodeToFetchDelay(params.decodeToFetchDelay),
+      commitToFetchDelay(params.commitToFetchDelay),
+      bacToFetchDelay(params.bacToFetchDelay),
+      fetchTargetWidth(params.fetchTargetWidth),
+      minInstSize(params.minInstSize),
+      numThreads(params.numThreads),
+      stats(_cpu,this)
+{
+    fatal_if(decoupledFrontEnd && (fetchTargetWidth < params.fetchBufferSize),
+            "Fetch target width should be larger than fetch buffer size!");
+
+    for (int i = 0; i < MaxThreads; i++) {
+        bacPC[i].reset(params.isa[0]->newPCState());
+        stalls[i] = {false, false, false};
+    }
+
+    assert(bpu!=nullptr);
```

Comment:
```suggestion
    assert(bpu != nullptr);
```

This PR may be a good test for the clang format...

Reply:
I run the clang formatter now on it.